v12/storage-dotnet-redundancy-ha-ra-grs/Program.cs

```diff
@@ -0,0 +1,262 @@
+                {
+                    Response<BlobDownloadInfo> response;
+
+                    Console.WriteLine($"\nDownloading blob {blobName}:");
+
+                    // If the number of retries exceeds the threshold, switch to the BlobServiceClient
+                    // pointing at the secondary region endpoint for read requests
+                    if (retryCount > retryThreshold)
+                    {
+                        response = await blobSecondary.DownloadAsync();
+                        secondaryReadCount++;
+
+                        // If the number of read requests against the secondary region endpoint exceeds
+                        // the threshold, reset the variables and try the primary region endpoint again
+                        if (secondaryReadCount > secondaryReadThreshold)
+                        {
+                            retryCount = 0;
+                            secondaryReadCount = 0;
+                        }
+                    }
+                    else
+                    {
+                        response = await blobClient.DownloadAsync();
+                    }
+
+                    BlobDownloadInfo downloadInfo = response.Value;
+
+                    // Write out the response status
+                    Console.WriteLine($"Response status: {response.GetRawResponse().Status} ({response.GetRawResponse().ReasonPhrase})");
+
+                    // Write out the blob data
+                    Console.Write("Blob data: ");
+                    Console.WriteLine((await BinaryData.FromStreamAsync(downloadInfo.Content)).ToString());
+                    Console.WriteLine($"\nRetry count: {retryCount}");
+                    Console.WriteLine($"Secondary region read count: {secondaryReadCount}");
+
+                    Console.WriteLine("\nPress any key to download the blob again - Esc to exit");
+                }
+            }
+            catch (RequestFailedException e)
+            {
+                Console.WriteLine(e.Message);
+                Console.ReadLine();
+                throw;
+            }
+            finally
+            {
+                Console.WriteLine("\nThe program has completed successfully");
+                Console.WriteLine("Press 'Enter' to delete the sample container and exit the application");
+                Console.ReadLine();
+
+                // Clean up resources
+                try
+                {
+                    if (containerClient != null)
+                    {
+                        Console.WriteLine($"Deleting the container {containerClient.Name}");
+                        await containerClient.DeleteAsync();
+                    }
+                }
+                catch (RequestFailedException e)
+                {
+                    Console.WriteLine(e.Message);
+                    Console.ReadLine();
+                }
+            }
+        }
+    }
+}
```

Comment:
After reading through all of this code, I see what it's doing, but it might be more clear to break this down a bit more. Have one container be the producer and the other be the consumer, and have them in separate files. The `Program.cs` file could serve as just the host that calls both into action and waits for either cancellation or completion. Instantiate one that starts producing, and then the other that starts consuming.